### PR TITLE
[flang] fix private pointers and default initialized variables

### PR DESCRIPTION
**Pull request by jeanPerier**
Both OpenMP privatization and DO CONCURRENT LOCAL lowering was incorrect for pointers and derived type with default initialization.

For pointers, the descriptor was not established with the rank/type code/element size, leading to undefined behavior if any inquiry was made to it prior to a pointer assignment (and if/when using the runtime for pointer assignments, the descriptor must have been established).

For derived type with default initialization, the copies were not default initialized.

### CODE REVIEW

File: flang/test/Lower/do_concurrent_local_default_init.f90 [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    end subroutine
+  end interface
+  character(*), pointer :: p(:)
+  integer :: i
+  do concurrent (i=1:10) local(p)
+    call takes_ptr(p)
+  end do
+end subroutine
+
+subroutine test_default_init(p)
```
Comment (eugeneepshteyn) [inline flang/test/Lower/do_concurrent_local_default_init.f90:17]: `p` is unused here, right? I don't think it's needed for this particular test, so maybe just remove `p`?

File: flang/test/Lower/OpenMP/delayed-privatization-default-init.f90 [hunk 2]
```diff
@@ -0,0 +1,27 @@
+! Test delayed privatization for derived types with default initialization.
+
+! RUN: %flang_fc1 -emit-hlfir -fopenmp -mmlir --openmp-enable-delayed-privatization \
+! RUN:   -o - %s 2>&1 | FileCheck %s
+! RUN: bbc -emit-hlfir -fopenmp --openmp-enable-delayed-privatization -o - %s 2>&1 |\
```
Comment (eugeneepshteyn) [inline flang/test/Lower/OpenMP/delayed-privatization-default-init.f90:5]: For my education, what does `bbc` tool do?

Comment (Leporacanthicus) [reply flang/test/Lower/OpenMP/delayed-privatization-default-init.f90:5]: `bbc` is an alternative frontend of the compiler, which generates the FIR (or HLFIR in this case) level. There is a `tco` second part that runs optimisation passes and generated LLVM-IR (or lowered MLIR). 

The main point of having both here is to check that the option for delayed privatization works for both variants. Almost all code is shared, but command-line parsing is not identical. [We should, as a general case, avoid running everything in both `bbc` and `flang`, if there's no clear reason why code takes a different path through the two tools].

File: flang/lib/Lower/Bridge.cpp [hunk 3]
```diff
@@ -822,7 +830,20 @@ class FirConverter : public Fortran::lower::AbstractConverter {
...
+          // with them.
+          if (const Fortran::semantics::DeclTypeSpec *declTypeSpec =
+                  sym.GetType())
+            if (const Fortran::semantics::DerivedTypeSpec *derivedTypeSpec =
+                    declTypeSpec->AsDerived())
+              if (derivedTypeSpec->HasDefaultInitialization(
+                      /*ignoreAllocatable=*/false, /*ignorePointer=*/false)) {
+                mlir::Value box = builder->createBox(loc, exv);
+                fir::runtime::genDerivedTypeInitialize(*builder, loc, box);
+              }
```
Comment (luporl) [inline flang/lib/Lower/Bridge.cpp:846]: `DataSharingProcessor::cloneSymbol` does something similar for privatizations, except for firstprivate:

```c++
  if (!isFirstPrivate &&
      Fortran::lower::hasDefaultInitialization(sym->GetUltimate()))
    Fortran::lower::defaultInitializeAtRuntime(converter, *sym, *symTable);
```

Won't this make `Initialize` be called twice for private clauses?

Comment (jeanPerier) [reply flang/lib/Lower/Bridge.cpp:848]: Indeed, thanks for catching this! I removed the init from `hasDefaultInitialization` and added a flag to the `createHostAssociateVarClone` to disable the init in the firstprivate case (and the doc concurrent local_init).

File: flang/test/Lower/do_concurrent_local_default_init.f90 [hunk 4]
```diff
@@ -0,0 +1,52 @@
+! Test default initialization of DO CONCURRENT LOCAL() entities.
+! RUN: bbc -emit-hlfir -I nowhere -o - %s | FileCheck %s
```
Comment (vdonaldson) [inline flang/test/Lower/do_concurrent_local_default_init.f90:2]: Is `-I nowhere` needed?

File: flang/lib/Lower/Bridge.cpp [hunk 5]
```diff
@@ -768,13 +768,21 @@ class FirConverter : public Fortran::lower::AbstractConverter {
     // Initialise cloned allocatable
     hexv.match(
         [&](const fir::MutableBoxValue &box) -> void {
-          // Do not process pointers
+          const auto new_box = exv.getBoxOf<fir::MutableBoxValue>();
           if (Fortran::semantics::IsPointer(sym.GetUltimate())) {
+            // Establish the pointer descriptors. The rank and type code/size
+            // at least must be set properly for later inquiry of the pointer
+            // to work, and new pointers are always given diassociated status
```
Comment (vdonaldson) [inline flang/lib/Lower/Bridge.cpp:775]: `diassociated` -> `disassociated`

**Review by vdonaldson — APPROVED** [review]
Thanks Jean

File: flang/test/Lower/do_concurrent_local_default_init.f90 (hunk 4 above)
Comment (jeanPerier) [reply flang/test/Lower/do_concurrent_local_default_init.f90:2]: With bbc it hides the intrinsic modules, which prevents generation of runtime type info globals for derived type, making the test faster to run and a lot easier to debug if needed.

File: flang/lib/Lower/Bridge.cpp (hunk 5 above)
Comment (jeanPerier) [inline flang/lib/Lower/Bridge.cpp:775]:
```suggestion
            // to work, and new pointers are always given disassociated status
```

**Comment by kiranchandramohan** [comment]
Thanks @jeanPerier for the fix.

**Pull request merged**